pkg/aggregateeventrecorder/aggregateeventrecorder.go

```diff
@@ -0,0 +1,172 @@
+				id := event.RetrievalId().String()
+
+				switch event.Code() {
+				case types.StartedCode:
+					// Save the start time so we can calculate time to first byte
+					eventTempMap[id] = tempData{event.Time(), 0}
+
+				case types.FirstByteCode:
+					// Calculate time to first byte
+					tempData := eventTempMap[id]
+					tempData.ttfb = event.Time().Sub(tempData.startTime).Seconds()
+					eventTempMap[id] = tempData
+
+				case types.SuccessCode, types.FailedCode:
+					tempData := eventTempMap[id]
+
+					success := event.Code() == types.SuccessCode
+					spId := event.StorageProviderId().String()
+					endTime := event.Time()
+
+					var bytesPerSec float64
+					if success {
+						byts := event.(events.RetrievalEventSuccess).ReceivedSize()
+						duration := event.(events.RetrievalEventSuccess).Duration().Seconds()
+						bytesPerSec = float64(byts) / duration
+					}
+
+					aggregatedEvent := AggregatedEvent{
+						InstanceID:        instanceID,
+						RetrievalID:       id,
+						StorageProviderID: spId,
+						TimeToFirstByte:   tempData.ttfb,
+						Bandwidth:         bytesPerSec,
+						Success:           success,
+						StartTime:         tempData.startTime,
+						EndTime:           endTime,
+					}
+
+					batchedData = append(batchedData, aggregatedEvent)
+					if emptyGaurdChan == nil {
+						emptyGaurdChan = postChan // emptyGuardChan is no longer nil when we add to the batch
+					}
+				}
+			case emptyGaurdChan <- batchedData: // won't execute while emptyGaurdChan is nil
+				batchedData = nil
+				emptyGaurdChan = nil
+			}
+		}
+	}()
+
+	// Start parallel http posters
+	for i := 0; i < parallelPosters; i++ {
+		go func() {
+			client := http.Client{Timeout: httpTimeout}
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case batchedData := <-postChan:
+
+					byts, err := json.Marshal(batchedEvents{batchedData})
+					if err != nil {
+						logging.Errorf("Failed to JSONify and encode event: %w", err.Error())
+						return
+					}
+
+					req, err := http.NewRequest("POST", endpointURL, bytes.NewBufferString(string(byts)))
+					if err != nil {
+						logging.Errorf("Failed to create POST request for [%s]: %w", endpointURL, err.Error())
+						return
+					}
+
+					req.Header.Set("Content-Type", "application/json")
+
+					// set authorization header if configured
+					if endpointAuthorization != "" {
+						req.Header.Set("Authorization", fmt.Sprintf("Basic %s", endpointAuthorization))
+					}
+
+					resp, err := client.Do(req)
+					if err != nil {
+						logging.Errorf("Failed to POST event to [%s]: %w", endpointURL, err.Error())
+						return
+					}
+
+					defer resp.Body.Close() // error not so important at this point
+					if resp.StatusCode < 200 || resp.StatusCode > 299 {
+						logging.Errorf("Expected success response code from server, got: %s", http.StatusText(resp.StatusCode))
+					}
+				}
+			}
+		}()
+	}
+
+	return func(event types.RetrievalEvent) {
```

Comment:
I find returning functions like this confusing.  This is a very large closure. Shouldn't we just make this a struct?